js/utils.js

```diff
@@ -1511,6 +1511,62 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
     });
   }
 
+  function setFilterValues(options) {
+    return new Promise(function(resolve) {
```

Comment:
You don't need to start a `new Promise()` here.

At the bottom of the function you could just use `return Promise.all(promises)`

Reply:
@tonytlwu,
as I understand it, the new Promise wrapper is needed for the `getCachedSession` function because we are waiting for a callback from the function, if we remove the wrapper, then the main function is not executed due to the fact that the callback does not have time to complete and the function occurs asynchronously.
Here is a vide without `new Promise` wrapper
http://somup.com/cYl32H2HSV

Reply:
@AndrRyaz I disagree. Tell me if the following structure misses any important race conditions.

```js
function setFilterValues(options) {
  options = options || {};

  if (!options.config) {
    return Promise.resolve();
  }

  // Using lodash function so the IF clause above doesn't need to check for options.config.filterOptions
  var promises = _.map(options.config.filterOptions, function() {
    return new Promise(function(resolve) {
      switch () {
        // Simplified summary of your current code (excl. other recommended changes)
        case 'use-profile-values':
          Fliplet.User.getCachedSession()
            .then(function(session) {
              item.value = data.entries.dataSource.data[item.fieldValue];
              resolve();
            });
          break;
        // ... 
      }
    });
  });

  // setFilterValues returns a promise that doesn't resolve until all promises have resolved
  return Promise.all(promises);
}
```